### 05. 2011 Machine Learning Dataset Preparation for Temporal Comparative Analysis

### Section-1. Objective

The objective of this notebook is to prepare a machine learning-ready dataset from the 2011 Toronto Neighbourhood Profiles for temporal comparative analysis.

The notebook identifies and extracts socioeconomic indicators that are comparable with the 2016 machine learning-ready dataset, transforms the original profile-style dataset into a neighbourhood-level analytical dataset, performs data quality assessment, and exports the cleaned dataset for use in Notebook 5.

#### Input

- `neighbourhood-profiles-2011-140-model.csv`

#### Output

- `neighbourhood_profiles_2011_ml_ready.csv`

#### Main Tasks

- Load the raw 2011 Toronto Neighbourhood Profiles dataset.
- Validate the dataset structure and quality.
- Identify and extract the selected socioeconomic indicators.
- Construct a machine learning-ready neighbourhood-level dataset.
- Perform data quality assessment.
- Export the processed dataset for temporal comparative analysis.

In [1]:
# Import libraries

import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [3]:
# Load the raw dataset

raw_df = pd.read_csv(
    "../../data/raw/neighbourhood-profiles-2011-140-model.csv")

print("Dataset Shape:", raw_df.shape)

display(raw_df.head())

Dataset Shape: (1537, 145)


,_id,Category,Topic,Attribute,City of Toronto,Agincourt North,Agincourt South-Malvern West,Alderwood,Annex,Banbury-Don Mills,Bathurst Manor,Bay Street Corridor,Bayview Village,Bayview Woods-Steeles,Bedford Park-Nortown,Beechborough-Greenbrook,Bendale,Birchcliffe-Cliffside,Black Creek,Blake-Jones,Briar Hill-Belgravia,Bridle Path-Sunnybrook-York Mills,Broadview North,Brookhaven-Amesbury,Cabbagetown-South St. James Town,Caledonia-Fairbank,Casa Loma,Centennial Scarborough,Church-Yonge Corridor,Clairlea-Birchmount,Clanton Park,Cliffcrest,Corso Italia-Davenport,Danforth,Danforth East York,Don Valley Village,Dorset Park,Dovercourt-Wallace Emerson-Junction,Downsview-Roding-CFB,Dufferin Grove,East End-Danforth,Edenbridge-Humber Valley,Eglinton East,Elms-Old Rexdale,Englemount-Lawrence,Eringate-Centennial-West Deane,Etobicoke West Mall,Flemingdon Park,Forest Hill North,Forest Hill South,Glenfield-Jane Heights,Greenwood-Coxwell,Guildwood,Henry Farm,High Park North,High Park-Swansea,Highland Creek,Hillcrest Village,Humber Heights-Westmount,Humber Summit,Humbermede,Humewood-Cedarvale,Ionview,Islington-City Centre West,Junction Area,Keelesdale-Eglinton West,Kennedy Park,Kensington-Chinatown,Kingsview Village-The Westway,Kingsway South,Lambton Baby Point,L'Amoreaux,Lansing-Westgate,Lawrence Park North,Lawrence Park South,Leaside-Bennington,Little Portugal,Long Branch,Malvern,Maple Leaf,Markland Wood,Milliken,Mimico (includes Humber Bay Shores),Morningside,Moss Park,Mount Dennis,Mount Olive-Silverstone-Jamestown,Mount Pleasant East,Mount Pleasant West,New Toronto,Newtonbrook East,Newtonbrook West,Niagara,North Riverdale,North St. James Town,Oakridge,Oakwood Village,O'Connor-Parkview,Old East York,Palmerston-Little Italy,Parkwoods-Donalda,Pelmo Park-Humberlea,Playter Estates-Danforth,Pleasant View,Princess-Rosethorn,Regent Park,Rexdale-Kipling,Rockcliffe-Smythe,Roncesvalles,Rosedale-Moore Park,Rouge,Runnymede-Bloor West Village,Rustic,Scarborough Village,South Parkdale,South Riverdale,St.Andrew-Windfields,Steeles,Stonegate-Queensway,Tam O'Shanter-Sullivan,Taylor-Massey,The Beaches,Thistletown-Beaumond Heights,Thorncliffe Park,Trinity-Bellwoods,University,Victoria Village,Waterfront Communities-The Island,West Hill,West Humber-Clairville,Westminster-Branson,Weston,Weston-Pelham Park,Wexford/Maryvale,Willowdale East,Willowdale West,Willowridge-Martingrove-Richview,Woburn,Woodbine Corridor,Woodbine-Lumsden,Wychwood,Yonge-Eglinton,Yonge-St.Clair,York University Heights,Yorkdale-Glen Park
0,1,Population,Population,"Population, 2011",2615060.0,30279.0,21988.0,11904.0,29177.0,26918.0,15434.0,19348.0,17671.0,13530.0,23185.0,6488.0,27876.0,21856.0,22057.0,7763.0,14302.0,8713.0,11563.0,17787.0,12053.0,9851.0,10487.0,13093.0,28349.0,24770.0,14612.0,15703.0,13743.0,9444.0,16712.0,26739.0,24363.0,34631.0,34659.0,11449.0,20839.0,14943.0,22829.0,9550.0,22086.0,18810.0,10927.0,22168.0,12474.0,10926.0,31390.0,14083.0,9816.0,11333.0,21292.0,21740.0,13097.0,17656.0,10583.0,12525.0,15853.0,14108.0,13091.0,38084.0,14027.0,10638.0,17058.0,18495.0,21723.0,9170.0,7921.0,44919.0,14642.0,14541.0,15070.0,17011.0,12050.0,9632.0,45086.0,10197.0,10436.0,27167.0,26541.0,17587.0,16306.0,13145.0,32788.0,15982.0,28593.0,10900.0,16423.0,23052.0,21274.0,12191.0,17832.0,13497.0,21073.0,18316.0,9118.0,13746.0,34617.0,8710.0,7653.0,16144.0,11197.0,10007.0,10488.0,22267.0,15050.0,20631.0,45912.0,9632.0,9951.0,16609.0,21251.0,25642.0,17958.0,25017.0,24691.0,27398.0,15594.0,21130.0,10138.0,19225.0,16802.0,7782.0,17182.0,43361.0,26547.0,34100.0,25446.0,18170.0,12010.0,27018.0,45041.0,15004.0,21343.0,53350.0,11703.0,7826.0,13986.0,10578.0,11652.0,27713.0,14687.0
1,2,Population,Population,"Population, 2006",2503281.0,30156.0,21562.0,11656.0,27482.0,25439.0,14945.0,15325.0,15376.0,13298.0,21961.0,6526.0,25479.0,21652.0,21720.0,7765.0,14198.0,8205.0,11596.0,17318.0,11120.0,10281.0,9815.0,12450.0,24379.0,19850.0,13732.0,14531.0,14405.0,9576.0,16500.0,26644.0,24374.0,34737.0,32010.0,11811.0,

### Section-2 Data overview

Inspect the structure of the raw 2011 Toronto Neighbourhood Profiles dataset by examining its dimensions, column names, data types, and missing values. This assessment provides an overview of the dataset and identifies any structural characteristics that should be considered during feature extraction and preprocessing.

In [32]:
# Dataset overview

print("=" * 60)
print("Dataset Information")
print("=" * 60)

print(f"Rows: {raw_df.shape[0]}")
print(f"Columns: {raw_df.shape[1]}")

print("\nData Types")
display(raw_df.dtypes.value_counts().to_frame("Count"))

print("\nFirst 10 Columns")
display( pd.DataFrame(
        raw_df.columns[:10],
        columns=["Column Name"]    ))

Dataset Information
Rows: 1537
Columns: 145

Data Types


,Count
float64,141
object,3
int64,1



First 10 Columns


,Column Name
0,_id
1,Category
2,Topic
3,Attribute
4,City of Toronto
5,Agincourt North
6,Agincourt South-Malvern West
7,Alderwood
8,Annex
9,Banbury-Don Mills


In [34]:
# Missing Values Summary

missing_summary = raw_df.isnull().sum()

print("\nMissing Values Summary")
print(f"Columns with missing values: {(missing_summary > 0).sum()}")

print(
    f"Maximum missing values in a single column: "
    f"{missing_summary.max()}")

print(
    f"Total missing values: {missing_summary.sum()}")


Missing Values Summary
Columns with missing values: 140
Maximum missing values in a single column: 9
Total missing values: 1260


### Interpretation

The initial inspection confirmed that the raw 2011 Toronto Neighbourhood Profiles dataset contains **1,537 socioeconomic indicators** across **145 variables**. The dataset follows the expected profile structure, where socioeconomic indicators are stored as rows and neighbourhoods are represented as columns.

The dataset consists primarily of numeric variables, with a small number of descriptive fields used for metadata. A limited number of missing values were identified, reflecting indicators that are not applicable to every neighbourhood. These will be addressed during feature extraction and data preparation.

### Section-3 Research Variable Identification

Identify the socioeconomic indicators required for the temporal comparative analysis. The selected variables were chosen based on their relevance to neighbourhood socioeconomic vulnerability and their compatibility with the 2016 machine learning-ready dataset developed in Notebook 1.

This step establishes the analytical variables that will be extracted from the raw 2011 Toronto Neighbourhood Profiles and transformed into the machine learning-ready dataset.

In [9]:
# Research variables for temporal comparative analysis

research_variables = [
    "Population",
    "Participation Rate",
    "Employment Rate",
    "Unemployment Rate",
    "Bachelor",
    "No certificate",
    "Visible minority",
    "Renter",
    "Immigrant",
    "Lone-parent",
    "Low Income Measure After Tax"]

research_df = pd.DataFrame(
    {"Selected Research Variables": research_variables})

display(research_df)

,Selected Research Variables
0,Population
1,Participation Rate
2,Employment Rate
3,Unemployment Rate
4,Bachelor
5,No certificate
6,Visible minority
7,Renter
8,Immigrant
9,Lone-parent


### Interpretation

The research variables were selected to support temporal comparison between the 2011 and 2016 Toronto Neighbourhood Profiles. These indicators represent the demographic, educational, labour market, housing, and socioeconomic characteristics identified as relevant to neighbourhood-level socioeconomic vulnerability.

The selected variables also maximize consistency with the 2016 machine learning-ready dataset, providing a common analytical framework for subsequent comparative analyses.

### Section- 4 Locate Research Variables

Locate the selected socioeconomic indicators within the raw 2011 Toronto Neighbourhood Profiles dataset and record their corresponding row indices. This step establishes the mapping between the research variables and their locations in the raw dataset, providing a transparent and reproducible foundation for feature extraction.

In [11]:
# Locate research variables in the raw dataset

search_terms = [
    "Population, 2011",
    "Participation rate",
    "Employment rate",
    "Unemployment rate",
    "Bachelor",
    "No certificate",
    "Visible minority",
    "Renter",
    "Immigrant",
    "Lone-parent",
    "Low Income Measure After Tax"]

for term in search_terms:
    print("=" * 60)
    print(f"Search Term: {term}")
    print("=" * 60)

    display(
        raw_df[
            raw_df["Attribute"].str.contains(
                term,
                case=False,
                na=False            )
        ][["_id", "Category", "Topic", "Attribute"]]    )

Search Term: Population, 2011


,_id,Category,Topic,Attribute
0,1,Population,Population,"Population, 2011"


Search Term: Participation rate


,_id,Category,Topic,Attribute
1412,1413,Labour,Labour force status,Participation rate


Search Term: Employment rate


,_id,Category,Topic,Attribute
1413,1414,Labour,Labour force status,Employment rate
1414,1415,Labour,Labour force status,Unemployment rate


Search Term: Unemployment rate


,_id,Category,Topic,Attribute
1414,1415,Labour,Labour force status,Unemployment rate


Search Term: Bachelor


,_id,Category,Topic,Attribute
1372,1373,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1373,1374,Education,"Highest certificate, diploma or degree","University certificate, diploma or degree ..."
1374,1375,Education,"Highest certificate, diploma or degree",Bachelor's degree
1375,1376,Education,"Highest certificate, diploma or degree","University certificate, diploma or degre..."
1382,1383,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1383,1384,Education,"Highest certificate, diploma or degree","University certificate, diploma or degree ..."
1384,1385,Education,"Highest certificate, diploma or degree",Bachelor's degree
1385,1386,Education,"Highest certificate, diploma or degree","University certificate, diploma or degre..."


Search Term: No certificate


,_id,Category,Topic,Attribute
1367,1368,Education,"Highest certificate, diploma or degree","No certificate, diploma or degree"
1377,1378,Education,"Highest certificate, diploma or degree","No certificate, diploma or degree"


Search Term: Visible minority


,_id,Category,Topic,Attribute
995,996,Ethnocultural diversity,Visible minority population,Total population in private households by visi...
996,997,Ethnocultural diversity,Visible minority population,Total visible minority population
1007,1008,Ethnocultural diversity,Visible minority population,"Visible minority, n.i.e."
1009,1010,Ethnocultural diversity,Visible minority population,Not a visible minority


Search Term: Renter


,_id,Category,Topic,Attribute
1305,1306,Housing,Housing tenure,Renter


Search Term: Immigrant


,_id,Category,Topic,Attribute
831,832,Immigration and citizenship,Immigrant status and period of immigration,Total population in private households by immi...
832,833,Immigration and citizenship,Immigrant status and period of immigration,Non-immigrants
833,834,Immigration and citizenship,Immigrant status and period of immigration,Immigrants
842,843,Immigration and citizenship,Age at immigration,Total immigrant population in private househol...
848,849,Immigration and citizenship,Immigrants by selected place of birth,Total population in private households by immi...
849,850,Immigration and citizenship,Immigrants by selected place of birth,Non-immigrants
852,853,Immigration and citizenship,Immigrants by selected place of birth,Immigrants
914,915,Immigration and citizenship,Recent immigrants by selected place of birth,Total recent immigrant population in private h...


Search Term: Lone-parent


,_id,Category,Topic,Attribute
153,154,"Families, households and marital status",Household type,Lone-parent family households
159,160,"Families, households and marital status",Household type,Lone-parent family households
193,194,"Families, households and marital status",Family structure,Total lone-parent families by sex of parent ...
787,788,Income,Income of families,Lone-parent economic families


Search Term: Low Income Measure After Tax


,_id,Category,Topic,Attribute


### Section-5 Variable Selection and Justification

Review the candidate indicators identified in the raw dataset and select the most appropriate variable for each research construct. The selected indicators are based on conceptual relevance, consistency with the 2016 machine learning-ready dataset, and suitability for temporal comparative analysis.

In [12]:
# Final research variable selection

variable_selection = pd.DataFrame({

    "Research Variable": [

        "Population",
        "Participation_Rate",
        "Employment_Rate",
        "Unemployment_Rate",
        "Bachelors_Degree",
        "No_Diploma",
        "Visible_Minority",
        "Renters",
        "Immigrant_Population",
        "Lone_Parent_Families",
        "Low_Income_AfterTax_Pct"    ],

    "Selected Indicator": [

        "Population, 2011",
        "Participation rate",
        "Employment rate",
        "Unemployment rate",
        "Bachelor's degree",
        "No certificate, diploma or degree",
        "Total visible minority population",
        "Renter",
        "Immigrants",
        "Total lone-parent families",
        "Low-income measure after tax (LIM-AT)"    ],

    "Reason for Selection": [

        "Represents total neighbourhood population.",
        "Matches the 2016 labour force participation indicator.",
        "Matches the 2016 employment indicator.",
        "Matches the 2016 unemployment indicator.",
        "Represents the population whose highest educational attainment is a bachelor's degree.",
        "Represents residents without formal educational qualifications.",
        "Represents the total visible minority population at the neighbourhood level.",
        "Represents renter households and aligns with the 2016 housing tenure indicator.",
        "Represents the total immigrant population used in the 2016 analysis.",
        "Represents total lone-parent families within each neighbourhood.",
        "Represents neighbourhood low-income prevalence and serves as the dependent variable throughout the analysis."]})

display(variable_selection)

,Research Variable,Selected Indicator,Reason for Selection
0,Population,"Population, 2011",Represents total neighbourhood population.
1,Participation_Rate,Participation rate,Matches the 2016 labour force participation in...
2,Employment_Rate,Employment rate,Matches the 2016 employment indicator.
3,Unemployment_Rate,Unemployment rate,Matches the 2016 unemployment indicator.
4,Bachelors_Degree,Bachelor's degree,Represents the population whose highest educat...
5,No_Diploma,"No certificate, diploma or degree",Represents residents without formal educationa...
6,Visible_Minority,Total visible minority population,Represents the total visible minority populati...
7,Renters,Renter,Represents renter households and aligns with t...
8,Immigrant_Population,Immigrants,Represents the total immigrant population used...
9,Lone_Parent_Families,Total lone-parent families,Represents total lone-parent families within e...


In [13]:
raw_df[
    raw_df["Attribute"].str.contains(
        "low",
        case=False,
        na=False
    )
][["_id", "Category", "Topic", "Attribute"]]

,_id,Category,Topic,Attribute
816,817,Income,Low income,In low income in 2010 based on after-tax low-i...
821,822,Income,Low income,Prevalence of low income in 2010 based on afte...
1372,1373,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1382,1383,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...


### Interpretation

The candidate indicators identified in the raw 2011 Toronto Neighbourhood Profiles dataset were systematically reviewed to select the most appropriate variable for each research construct. The selection process was guided by conceptual relevance, data availability, and consistency with the 2016 machine learning-ready dataset.

Where multiple candidate indicators were available, the indicator that most closely represented the corresponding socioeconomic concept was selected. This approach ensured that the final feature set provides a transparent, reproducible, and methodologically consistent foundation for temporal comparative analysis between the 2011 and 2016 datasets.

### Section-6. Final Feature Mapping

Create a final feature mapping that links each selected research variable to its corresponding indicator and source row within the raw 2011 Toronto Neighbourhood Profiles dataset. This mapping provides a transparent and reproducible reference for extracting the selected variables and constructing the machine learning-ready dataset.

In [15]:
# Final feature mapping

mapping_df = pd.DataFrame({

    "Research Variable": [

        "Population",
        "Participation_Rate",
        "Employment_Rate",
        "Unemployment_Rate",
        "Bachelors_Degree",
        "No_Diploma",
        "Visible_Minority",
        "Renters",
        "Immigrant_Population",
        "Lone_Parent_Families",
        "Low_Income_AfterTax_Pct"    ],

    "Selected Indicator": [

        "Population, 2011",
        "Participation rate",
        "Employment rate",
        "Unemployment rate",
        "Bachelor's degree",
        "No certificate, diploma or degree",
        "Total visible minority population",
        "Renter",
        "Immigrants",
        "Total lone-parent families",
        "Prevalence of low income in 2010 based on after-tax low-income measure (LIM-AT)"],

    "Source Row": [

        0,
        1412,
        1413,
        1414,
        1384,
        1367,
        996,
        1305,
        833,
        193,
        821    ]})

display(mapping_df)

,Research Variable,Selected Indicator,Source Row
0,Population,"Population, 2011",0
1,Participation_Rate,Participation rate,1412
2,Employment_Rate,Employment rate,1413
3,Unemployment_Rate,Unemployment rate,1414
4,Bachelors_Degree,Bachelor's degree,1384
5,No_Diploma,"No certificate, diploma or degree",1367
6,Visible_Minority,Total visible minority population,996
7,Renters,Renter,1305
8,Immigrant_Population,Immigrants,833
9,Lone_Parent_Families,Total lone-parent families,193


### Interpretation

The final feature mapping establishes a direct relationship between each selected research variable, its corresponding socioeconomic indicator, and its source row within the raw 2011 Toronto Neighbourhood Profiles dataset.

The mapped variables include the independent variables used to explain neighbourhood socioeconomic conditions and the dependent variable, **Low_Income_AfterTax_Pct**, which represents neighbourhood-level low-income prevalence. Documenting the selected indicator and source row for each variable ensures that the feature extraction process is transparent, reproducible, and methodologically consistent for constructing the 2011 machine learning-ready dataset.

### Section-7 Feature Extraction

Extract the selected socioeconomic indicators from the raw 2011 Toronto Neighbourhood Profiles dataset using the final feature mapping. This step creates a subset containing only the research variables required for constructing the machine learning-ready dataset.

In [22]:
# Extract selected variables

selected_df = raw_df.iloc[
    list(feature_mapping.values())].copy()

print("=" * 60)
print("Feature Extraction Summary")
print("=" * 60)

print(f"Selected Variables: {selected_df.shape[0]}")
print(f"Dataset Columns: {selected_df.shape[1]}")

display(selected_df[["_id", "Category", "Topic", "Attribute"] ])

Feature Extraction Summary
Selected Variables: 11
Dataset Columns: 145


,_id,Category,Topic,Attribute
0,1,Population,Population,"Population, 2011"
1412,1413,Labour,Labour force status,Participation rate
1413,1414,Labour,Labour force status,Employment rate
1414,1415,Labour,Labour force status,Unemployment rate
1384,1385,Education,"Highest certificate, diploma or degree",Bachelor's degree
1367,1368,Education,"Highest certificate, diploma or degree","No certificate, diploma or degree"
996,997,Ethnocultural diversity,Visible minority population,Total visible minority population
1305,1306,Housing,Housing tenure,Renter
833,834,Immigration and citizenship,Immigrant status and period of immigration,Immigrants
193,194,"Families, households and marital status",Family structure,Total lone-parent families by sex of parent ...


### Interpretation

The selected socioeconomic indicators were successfully extracted from the raw 2011 Toronto Neighbourhood Profiles dataset using the final feature mapping. The extracted subset contains **11 research variables** and preserves the original profile structure, where socioeconomic indicators remain as rows and neighbourhoods as columns.

Validation of the extracted records confirmed that each research variable corresponds to the intended socioeconomic indicator identified during the variable selection process. This verified subset provides the foundation for transforming the data into a neighbourhood-level machine learning-ready dataset for subsequent temporal comparative analysis.

### Section-8 Trasform data structure

Transform the extracted socioeconomic indicators from the original profile format into a neighbourhood-level analytical dataset. This process converts socioeconomic indicators from rows to columns, creating a machine learning-ready structure where each row represents a neighbourhood and each column represents a research variable.

In [23]:
# Transform the dataset into neighbourhood-level observations

ml_ready_df = (
    selected_df
    .drop(columns=["_id", "Category", "Topic", "Attribute"])
    .transpose())

ml_ready_df.columns = list(feature_mapping.keys())

ml_ready_df.index.name = "Neighbourhood"

ml_ready_df = ml_ready_df.reset_index()

print("=" * 60)
print("Dataset Transformation Summary")
print("=" * 60)

print(f"Rows (Neighbourhoods): {ml_ready_df.shape[0]}")
print(f"Columns (Variables): {ml_ready_df.shape[1]}")

display(ml_ready_df.head())

Dataset Transformation Summary
Rows (Neighbourhoods): 141
Columns (Variables): 12


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Visible_Minority,Renters,Immigrant_Population,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,City of Toronto,2615060.0,64.3,58.3,9.3,362580.0,380965.0,1264395.0,476085.0,1252215.0,146990.0,19.3
1,Agincourt North,30279.0,57.9,51.3,11.3,3530.0,5965.0,27025.0,1805.0,21490.0,1595.0,18.8
2,Agincourt South-Malvern West,21988.0,59.3,53.0,10.7,2520.0,4070.0,17780.0,1640.0,14225.0,1245.0,17.9
3,Alderwood,11904.0,66.0,61.1,7.4,1170.0,2290.0,1955.0,1025.0,4165.0,710.0,9.4
4,Annex,29177.0,71.8,66.6,7.3,6040.0,1690.0,6010.0,9780.0,8560.0,870.0,20.3


### Section-9 Remove the City of Toronto Aggregate Row

Remove the City of Toronto aggregate record from the transformed dataset. The aggregate represents a city-wide summary rather than an individual neighbourhood and is therefore excluded to ensure that all observations correspond to comparable neighbourhood-level units for temporal comparative analysis.

In [28]:
# Remove the City of Toronto aggregate row

ml_ready_df = (
    ml_ready_df[
        ml_ready_df["Neighbourhood"] != "City of Toronto" ].reset_index(drop=True))

print("City of Toronto Present:",
    "City of Toronto" in ml_ready_df["Neighbourhood"].values)

print("=" * 60)
print("Neighbourhood Dataset Summary")
print("=" * 60)

print(f"Neighbourhoods: {ml_ready_df.shape[0]}")
print(f"Research Variables: {ml_ready_df.shape[1] - 1}")
print("Identifier Column: 1")
print(f"Total Columns: {ml_ready_df.shape[1]}")

display(ml_ready_df.head())

City of Toronto Present: False
Neighbourhood Dataset Summary
Neighbourhoods: 140
Research Variables: 11
Identifier Column: 1
Total Columns: 12


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Visible_Minority,Renters,Immigrant_Population,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,Agincourt North,30279.0,57.9,51.3,11.3,3530.0,5965.0,27025.0,1805.0,21490.0,1595.0,18.8
1,Agincourt South-Malvern West,21988.0,59.3,53.0,10.7,2520.0,4070.0,17780.0,1640.0,14225.0,1245.0,17.9
2,Alderwood,11904.0,66.0,61.1,7.4,1170.0,2290.0,1955.0,1025.0,4165.0,710.0,9.4
3,Annex,29177.0,71.8,66.6,7.3,6040.0,1690.0,6010.0,9780.0,8560.0,870.0,20.3
4,Banbury-Don Mills,26918.0,61.3,57.0,7.1,4370.0,2300.0,11930.0,4730.0,12765.0,1165.0,12.8


### Interpretation

The City of Toronto aggregate record was successfully removed from the machine learning-ready dataset because it represents a city-wide summary rather than an individual neighbourhood. Including this aggregate observation could bias descriptive statistics and subsequent analyses, as it is not directly comparable to neighbourhood-level records.

The resulting dataset contains **140 neighbourhood-level observations**, **11 research variables** (10 independent variables and 1 dependent variable), and a single identifier column (`Neighbourhood`). This finalized structure provides a consistent analytical dataset for temporal comparison with the 2016 machine learning-ready dataset and for subsequent statistical analyses and predictive modelling.

### Section-10 Data Quality Assessment

Evaluate the quality of the machine learning-ready dataset by examining its structure, data types, missing values, and duplicate records. This assessment verifies that the transformed dataset is complete, internally consistent, and suitable for subsequent temporal comparative analysis.

In [30]:
# Data quality assessment

print("=" * 60)
print("Data Quality Summary")
print("=" * 60)

print(f"Dataset Shape: {ml_ready_df.shape}")

print("\nData Types")

dtype_summary = (
    ml_ready_df.dtypes
    .value_counts()
    .rename_axis("Data Type")
    .reset_index(name="Count"))

display(dtype_summary)

print("\nMissing Values")

missing_values = ml_ready_df.isnull().sum().sum()
print(f"Total Missing Values: {missing_values}")

print("\nDuplicate Records")

duplicates = ml_ready_df.duplicated().sum()
print(f"Duplicate Records: {duplicates}")

Data Quality Summary
Dataset Shape: (140, 12)

Data Types


,Data Type,Count
0,float64,11
1,object,1



Missing Values
Total Missing Values: 0

Duplicate Records
Duplicate Records: 0


### Interpration

The data quality assessment confirmed that the transformed dataset is complete, internally consistent, and suitable for subsequent analysis. The final machine learning-ready dataset contains **140 neighbourhood-level observations**, **11 numeric research variables**, and a single identifier column (`Neighbourhood`).

The validation identified **no missing values** and **no duplicate records**, indicating that the data preparation and transformation processes were successfully completed. The resulting dataset is fully prepared for temporal comparative analysis and can be used directly in the statistical analyses and modelling presented in Notebook 5.

### Section-11. Export the Machine Learning-Ready Dataset

Export the validated 2011 machine learning-ready dataset to a CSV file for use in subsequent temporal comparative analysis. This step preserves the final analytical dataset and provides a standardized input for Notebook 5.

In [38]:
# Export the machine learning-ready dataset

output_path = "../../data/processed/neighbourhood_profiles_2011_ml_ready.csv"

ml_ready_df.to_csv(
    output_path,
    index=False)

print("=" * 60)
print("Dataset Export Summary")
print("=" * 60)

print("Export completed successfully.")
print(f"Output File: {output_path}")
print(f"Dataset Shape: {ml_ready_df.shape}")

Dataset Export Summary
Export completed successfully.
Output File: ../../data/processed/neighbourhood_profiles_2011_ml_ready.csv
Dataset Shape: (140, 12)


### 12. Conclusion

This notebook prepared a machine learning-ready dataset from the 2011 Toronto Neighbourhood Profiles to support temporal comparative analysis with the 2016 dataset developed in Notebook 1.

A systematic feature engineering workflow was implemented to identify, justify, extract, transform, and validate the selected socioeconomic indicators. The original profile-style dataset was converted into a neighbourhood-level analytical dataset, and the City of Toronto aggregate record was removed to ensure that all observations represented comparable neighbourhood-level units.

The final dataset contains **140 neighbourhoods**, **11 research variables** (10 independent variables and 1 dependent variable), and **one identifier column** (`Neighbourhood`). Data quality assessment confirmed that the dataset contains no missing values or duplicate records, indicating that it is complete, internally consistent, and suitable for subsequent temporal comparative analysis.

The exported machine learning-ready dataset provides a transparent, reproducible, and methodologically consistent foundation for the comparative analyses presented in Notebook 5.